Configuration option to execute docker:stop in case of the failing build

### Description
We use the plugin the start some docker images in the `pre-integration-test` phase and stop them in the `post-integration-test`. This works very well. However, if any of the integration tests fails the build, then the created docker containers remain running. Thus, it would be great if there would be a configuration option to run `docker:stop` in case the build fails. This way everything would be cleaned up correctly.

This is a feature request, however I am not sure how hard it is to implement this or if it's possible at all.

This actually does not work in my case with the failsafe plugin. 

I hook the `flyway-maven-plugin` to the `integration-test` phase in order to verify flyway migrations against the mysql database version we use. However, if the flyway plugin fails the build, then `post-integration-test` is not called as expected. Here's my output:

```
...
[INFO] --- docker-maven-plugin:0.23.0:start (start-mysql) @ xyz ---
[INFO] DOCKER> [mysql:5.5.44] "mysql-flyway-test": Start container 901a6abd4d32
[INFO] DOCKER> [mysql:5.5.44] "mysql-flyway-test": Waiting for ports [3306] directly on container with IP (172.17.0.3).
[INFO] DOCKER> [mysql:5.5.44] "mysql-flyway-test": Waited on tcp port '[/172.17.0.3:3306]' 6545 ms
[INFO] 
[INFO] --- flyway-maven-plugin:3.2.1:migrate (default) @ xyz ---
[INFO] Flyway 3.2.1 by Boxfuse
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 40.937 s
[INFO] Finished at: 2018-01-10T11:45:28+01:00
[INFO] Final Memory: 67M/893M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.flywaydb:flyway-maven-plugin:3.2.1:migrate (default) on project xyz: org.flywaydb.core.api.FlywayException: Unable to obtain Jdbc connection from DataSource (jdbc:mysql://localhostsmth:13306/flyway?createDatabaseIfNotExist=true) for user 'root': Communications link failure
```

I am not sure if the setup of failsafe is correct, I simply copied the one from the official page:
```
   <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-failsafe-plugin</artifactId>
        <version>2.20.1</version>
        <executions>
          <execution>
            <goals>
              <goal>integration-test</goal>
              <goal>verify</goal>
            </goals>
          </execution>
        </executions>
      </plugin>
```
However, `post-integration-test` is not called. Any suggestions?

--- a/src/main/java/io/fabric8/maven/docker/StopMojo.java
+++ b/src/main/java/io/fabric8/maven/docker/StopMojo.java
@@ -11,6 +11,7 @@ import java.util.Set;
 import java.util.regex.Matcher;
 
 import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugin.MojoFailureException;
 import org.apache.maven.plugins.annotations.LifecyclePhase;
 import org.apache.maven.plugins.annotations.Mojo;
 import org.apache.maven.plugins.annotations.Parameter;
@@ -70,6 +71,41 @@ public class StopMojo extends AbstractDockerMojo {
 
     @Parameter(property = "docker.stopNamePattern")
     private String stopNamePattern;
+
+    /**
+     * If true, the containers are not stopped right away, but when the build is finished (success or failed).
+     */
+    @Parameter(property = "docker.executeStopOnVMShutdown", defaultValue = "false")
+    private boolean executeStopOnVMShutdown;
+
+    @Override
+    public void execute() throws MojoExecutionException, MojoFailureException {
+        if (this.executeStopOnVMShutdown) {
+            this.executeStopOnVMShutdown = false;
+           if (!invokedTogetherWithDockerStart()) {
+                throw new MojoExecutionException("docker:stop with executeStopOnVMShutdown is only possible if" +
+                        " the docker containers started within the same maven session.");
+            }
+
+            try {
+                // HttpDelete class is not used during start mojo, so we need to load it and initialize it. It is not
+                // possible to load classes in the shutdown hook as
+                Class.forName("org.apache.http.client.methods.HttpDelete", true, this.getClass().getClassLoader());
+            } catch (ClassNotFoundException e) {
+                e.printStackTrace();
+            }
+
+            Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+                try {
+                    super.execute();
+                } catch (MojoExecutionException | MojoFailureException e) {
+                    e.printStackTrace();
+                }
+            }));
+        } else {
+            super.execute();
+        }
+    }
 
     @Override
     protected void executeInternal(ServiceHub hub) throws MojoExecutionException, IOException, ExecException {

